analysis/src/Xform.ml

```diff
@@ -249,14 +271,144 @@ module AddTypeAnnotation = struct
-              ~uri:path ~newText ~range
+              ~edit:
+                {
+                  documentChanges =
+                    [
+                      TextDocumentEdit
+                        {
+                          textDocument = {version = None; uri};
+                          edits = [{newText; range}];
+                        };
+                    ];
+                }
           in
           codeActions := codeAction :: !codeActions
         | _ -> ()))
 end
 
+module TypeToModule = struct
+  (* Convert type definition into its own module *)
+  let mkIterator ~pos ~result ~newTypeName =
+    let changeTypeDecl (typ : Parsetree.type_declaration) ~txt =
+      let manifest = typ.ptype_manifest in
+      Ast_helper.Type.mk
+        {txt; loc = typ.ptype_name.loc}
+        ~loc:typ.ptype_loc ~attrs:typ.ptype_attributes ~params:typ.ptype_params
+        ~cstrs:typ.ptype_cstrs ~kind:typ.ptype_kind ~priv:typ.ptype_private
+        ?manifest
+    in
+
+    let structure_item (iterator : Ast_iterator.iterator)
+        (si : Parsetree.structure_item) =
+      match si.pstr_desc with
+      | Pstr_type (Nonrecursive, firstypeDec :: rest)
+        when Loc.hasPos ~pos si.pstr_loc ->
+        let loc = si.pstr_loc in
+        let firsLetterOfModule =
+          String.get firstypeDec.ptype_name.txt 0
+          |> Char.uppercase_ascii |> String.make 1
+        in
+        let restName =
+          String.sub firstypeDec.ptype_name.txt 1
+            (String.length firstypeDec.ptype_name.txt - 1)
+        in
+        let modName = firsLetterOfModule ^ restName in
+        let newFistType = changeTypeDecl firstypeDec ~txt:newTypeName in
+        let restStrucTypes =
+          Ast_helper.Str.type_ ~loc Nonrecursive ([newFistType] @ rest)
+        in
+        let pmb_expr = Ast_helper.Mod.structure ~loc [restStrucTypes] in
+        let mod_expr =
+          Parsetree.Pstr_module
+            {
+              pmb_name = {txt = modName; loc};
+              pmb_expr;
+              pmb_attributes = [];
+              pmb_loc = loc;
+            }
+        in
+        let processTypeKind (typeDec : Parsetree.type_declaration) =
+          match typeDec.ptype_kind with
+          | Ptype_record [{pld_name = {loc; txt}}; _] -> Some (loc, txt)
+          | Ptype_variant [{pcd_name = {loc; txt}}; _] -> Some (loc, txt)
+          | _ -> None
+        in
+        let references =
+          let referencesInfo =
+            [firstypeDec] @ rest |> List.filter_map processTypeKind
+          in
+          [(firstypeDec.ptype_name.loc, newTypeName)] @ referencesInfo
+        in
+        result := Some (Ast_helper.Str.mk ~loc mod_expr, references, modName);
+        Ast_iterator.default_iterator.structure_item iterator si
+      | _ -> ()
+    in
+    {Ast_iterator.default_iterator with structure_item}
+
+  let xform ~path ~pos ~codeActions ~printStructureItem structure ~debug =
+    let result = ref None in
+    let newTypeName = "t" in
```

Comment:
Feels like this logic should be factored out into its own easy to modify little local module.